### PR TITLE
[CALCITE] Handle ambiguous column resolution for unknown tables.

### PR DESCRIPTION
**Pull request by dstekol**
Modifies column resolution for unknown columns, so that unknown columns are not attributed to arbitrary unknown tables.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java [hunk 1]
```diff
@@ -212,7 +214,16 @@ public RelDataType resolveColumn(String columnName, SqlNode ctx) {
           nameMatcher.field(childRowType, columnName);
       if (field != null) {
         found++;
-        type = field.getType();
+        if (child.namespace.getType() instanceof UnknownRecordType) {
+          unknownFound++;
+          // If table is unknown, only set {@code type} to ANY if it has not already been set.
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:219]: nit: Line wrap

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java [hunk 2]
```diff
@@ -212,7 +214,16 @@ public RelDataType resolveColumn(String columnName, SqlNode ctx) {
...
+          // This ensures that the column will be resolved to a known (non-ANY) type if possible.
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:220]: nit: Line wrap

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java [hunk 3]
```diff
@@ -221,9 +232,15 @@ public RelDataType resolveColumn(String columnName, SqlNode ctx) {
     case 1:
       return type;
     default:
+      // Only return type if unknown tables are allowed, and at most one of the matched tables
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:235]: nit: Line wrap at 80 characters for this comment line and the rest in this block.

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java [hunk 4]
```diff
@@ -221,9 +232,15 @@ public RelDataType resolveColumn(String columnName, SqlNode ctx) {
...
+      // is a known table (if more than one known table is matched, then the column would be
+      // ambiguous even with full schema information). If more than one unknown table is matched,
+      // it doesn't matter which one the column actually comes from, since the column type is ANY.
+      if (validator.config().allowUnknownTables() && found - unknownFound <= 1) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:239]: nit: Line wrap

Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:239]: Wouldn't this match if, say, found == 2 and unknownFound == 2? Do we want that to match?

Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:239]: I think we would - that would mean that both of the matched tables are unknown, which is fine. The only case we want to throw an error is when more than one known table is matched.

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java (hunk 3 above)
Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:235]: Oops, accidentally set wrapping to 100.

File: core/src/main/java/org/apache/calcite/sql/validate/ListScope.java (hunk 4 above)
Comment (dasch-1) [reply core/src/main/java/org/apache/calcite/sql/validate/ListScope.java:239]: SGTM

File: core/src/test/java/org/apache/calcite/test/SqlValidatorBestEffortTest.java [hunk 5]
```diff
@@ -143,4 +143,51 @@
...
+        + " ON `EMP`.`EMPNO` = `FOO`.`X`";
+    sql(sql)
+        .withValidatorIdentifierExpansion(true)
+        .rewritesTo(expected);
+  }
+
+  @Test public void testColumnFromKnownTableExpanded() {
+    String sql = "SELECT empno "
+        + "FROM foo "
+        + "INNER JOIN emp on emp.empno = foo.x";
```
Comment (dasch-1) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorBestEffortTest.java:165]: Out of curiosity, where are "emp" and "dept" defined as known tables?

Comment (dstekol) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorBestEffortTest.java:165]: [MockCatalogReaderSimple](https://github.com/dstekol/calcite/blob/paren_from_clause/core/src/test/java/org/apache/calcite/test/catalog/MockCatalogReaderSimple.java)
Also linked in design doc in test section.

Comment (dasch-1) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorBestEffortTest.java:165]: Great thanks

**Pull request merged**